src/coreclr/jit/emit.h

```diff
@@ -2288,31 +2292,134 @@ class emitter
         return (emitCurIG && emitCurIGfreeNext > emitCurIGfreeBase);
     }
 
+#define EMIT_MAX_IG_INS_COUNT 256
+
+#if EMIT_BACKWARDS_NAVIGATION
+#define EMIT_MAX_PEEPHOLE_INS_COUNT 32 // The max number of previous instructions to navigate through for peepholes.
+#endif                                 // EMIT_BACKWARDS_NAVIGATION
+
     instrDesc* emitLastIns;
     insGroup*  emitLastInsIG;
 
 #if EMIT_BACKWARDS_NAVIGATION
     unsigned emitLastInsFullSize;
 #endif // EMIT_BACKWARDS_NAVIGATION
 
+    // Check to see if the last instruction is available.
+    inline bool emitHasLastIns() const
+    {
+        return (emitLastIns != nullptr);
+    }
+
+    // Checks to see if we can cross between the two given IG boundaries.
+    //
+    // We have the following checks:
+    // 1. Looking backwards across an IG boundary can only be done if we're in an extension IG.
+    // 2. The IG of the previous instruction must have the same GC interrupt status as the current IG.
+    //    This is related to #2; it disallows peephole when the previous IG is GC and the current is NOGC.
+    inline bool canPeepholeCrossIGBoundaries(insGroup* prevIG, insGroup* ig) const
+    {
+        if (emitCurIG == ig)
+        {
+            return (((emitCurIGinsCnt > 0) || (ig->igFlags & IGF_EXTEND)) && // We're not at the start of a new
+                                                                             // IG or we are at the start of a
+                                                                             // new IG, and it's an extension IG.
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        else if (prevIG != ig)
+        {
+            return ((prevIG->igFlags & IGF_EXTEND) &&
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        return true;
```

Comment:
Assuming you rename the arguments `prevInsIG` and `curInsIG` as suggested, I think this would be more clear:

```suggestion
        if (prevInsIG == curInsIG)
        {
             return true;
        }
        else 
        {
            return (curInsIG->igFlags & IGF_EXTEND) &&
                    ((prevInsIG->igFlags & IGF_NOGCINTERRUPT) == (curInsIG->igFlags & IGF_NOGCINTERRUPT));
        }
```